File: frontend/app/lib/models/user_log.dart
/// 사용자 행동 로그 한 건. 서버 user_log 테이블(POST /user-logs)과 1:1이다.
///
/// 화면은 이 클래스를 직접 만들지 않는다 — [writeLog]에 이름과 detail만 넘기면
/// 세션·플랫폼·앱 버전은 UserLogService가 채운다.
class UserLog {
  const UserLog({
    required this.name,
    required this.detail,
    required this.sessionId,
    required this.platform,
    required this.appVersion,
  });

  final LogName name;
  final Map<String, dynamic> detail;
  final String sessionId;
  final String platform;
  final String appVersion;

  Map<String, dynamic> toJson() => {
    'log_name': name.wire,
    'detail': detail,
    'session_id': sessionId,
    'platform': platform,
    'app_version': appVersion,
  };
}

/// 로그 이름. 서버 app/user_log.py의 LOG_NAMES와 같아야 한다 — 목록 밖 이름은
/// 서버가 422로 거절한다. 새 로그는 서버 목록에 먼저 넣고 여기 한 줄 추가한다.
enum LogName {
  // 세션
  appOpen('app_open'),
  // 홈
  homeCourseClick('home_course_click'),
  bannerClick('banner_click'),
  notificationOpen('notification_open'),
  // 코스 탐색
  coursePreviewOpen('course_preview_open'),
  courseSearch('course_search'),
  courseListSort('course_list_sort'),
  favoriteAdd('favorite_add'),
  favoriteRemove('favorite_remove'),
  navigateClick('navigate_click'),
  // 러닝
  runStart('run_start'),
  runPause('run_pause'),
  runResume('run_resume'),
  runFinish('run_finish'),
  runAbandon('run_abandon'),
  runUploadFailed('run_upload_failed'),
  runDetailOpen('run_detail_open'),
  // 스탬프·쿠폰
  stampDetailOpen('stamp_detail_open'),
  couponView('coupon_view'),
  couponUseClick('coupon_use_click'),
  // 커뮤니티
  externalLinkOpen('external_link_open'),
  // 계정
  login('login'),
  loginFailed('login_failed'),
  logout('logout'),
  withdraw('withdraw');

  const LogName(this.wire);

  /// 서버에 보내는 문자열.
  final String wire;
}

/// detail에 쓰는 키. 화면마다 표기가 갈리지 않게 여기 것만 쓴다(snake_case).
abstract final class LogKeys {
  static const courseId = 'course_id';
  static const runId = 'run_id';
  static const noticeId = 'notice_id';
  static const source = 'source';
  static const section = 'section';
  static const position = 'position';
  static const reason = 'reason';
  static const elapsedSec = 'elapsed_sec';
  static const distanceM = 'distance_m';
  static const error = 'error';
}

/// 어디서 눌렀는지(detail[LogKeys.source]).
enum LogSource { home, map, list, search, favorite, stamp, profile, detail }

File: server/app/user_log.py
"""user_log 공용 — 허용 로그 이름과 기록 헬퍼.

앱(POST /user-logs)과 서버 내부(코스 상세 조회 기록)가 같은 테이블에 쓰므로 로그 이름
목록을 여기서 한 번만 관리한다. 목록에 없는 이름은 거절해 오타·임의 문자열이
쌓이는 것을 막는다. 새 로그를 추가할 때는 이 목록에 먼저 넣는다.
"""

from sqlalchemy import insert
from sqlalchemy.exc import SQLAlchemyError
from sqlalchemy.orm import Session

from app.models import UserLog

# detail(JSONB) 한 건의 직렬화 상한(바이트). 경로·이미지 같은 큰 값을 막는다.
DETAIL_MAX_BYTES = 2048

# 서버가 직접 기록하는 로그.
COURSE_DETAIL_OPEN = "course_detail_open"

LOG_NAMES: frozenset[str] = frozenset(
    {
        COURSE_DETAIL_OPEN,
        # 홈
        "home_course_click",
        "banner_click",
        "notification_open",
        # 코스 탐색
        "course_preview_open",
        "course_search",
        "course_list_sort",
        "favorite_add",
        "favorite_remove",
        "navigate_click",
        # 러닝
        "run_start",
        "run_pause",
        "run_resume",
        "run_finish",
        "run_abandon",
        "run_upload_failed",
        "run_detail_open",
        "run_share",
        "run_share_failed",
        # 스탬프·쿠폰
        "stamp_detail_open",
        "coupon_view",
        "coupon_use_click",
        # 커뮤니티
        "external_link_open",
        # 계정
        "login",
        "login_failed",
        "logout",
        "withdraw",
        # 세션
        "app_open",
    }
)

PLATFORMS: frozenset[str] = frozenset({"android", "ios"})


def record_best_effort(
    db: Session,
    *,
    log_name: str,
    user_id: str | None,
    detail: dict,
    session_id: str | None = None,
    platform: str | None = None,
    app_version: str | None = None,
) -> None:
    """서버 내부 기록용(best-effort). 실패해도 롤백만 하고 호출한 요청은 성공시킨다.

    분석용 쓰기가 핵심 읽기(코스 상세 등)를 깨면 안 되기 때문이다. 세션·플랫폼·앱
    버전은 요청 헤더(deps.client_context)에서 온 값이고, 없으면 null로 남는다.
    """
    stmt = insert(UserLog).values(
        log_name=log_name,
        user_id=user_id,
        detail=detail,
        session_id=session_id,
        platform=platform,
        app_version=app_version,
    )
    try:
        db.execute(stmt)
        db.commit()
    except SQLAlchemyError:
        db.rollback()
